messaging/broker.go

```diff
@@ -818,6 +878,80 @@ func (t *Topic) WriteMessage(m *Message) error {
 	return nil
 }
 
+// Truncate attempts to delete topic segments such that the total size of the topic on-disk
+// is equal to or less-than maxSize. Returns the number of bytes deleted, and error if any.
+// This function is not guaranteed to be performant.
+func (t *Topic) Truncate(maxSize int64) (nBytesDeleted int64, err error) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	var segments Segments
+	segments, err = ReadSegments(t.Path())
+	if err != nil {
+		return
+	}
+
+	var totalSize int64
+	totalSize, err = segments.Size()
+	if err != nil {
+		return
+	}
+
+	// Find the highest-replicated index, this is a condition for deletion of segments
+	// within this topic.
+	var highestIndex uint64
+	for _, idx := range t.indexByURL {
+		if idx > highestIndex {
+			highestIndex = idx
+		}
+	}
+
+	var size int64
+	for {
+		if (totalSize - nBytesDeleted) <= maxSize {
+			break
+		}
+
+		// More than 2 segments are needed for current writes and replication checks.
+		if len(segments) <= 2 {
+			break
+		}
+
+		// Attempt deletion of oldest segment in topic. First and second segment needed
+		// for this operation.
+		first := segments[0]
+		second := segments[1]
+
+		// The first segment can only be deleted if the last index in that segment has
+		// been replicated. The most efficient way to check this is to ensure that the
+		// first index of the subsequent segment has been replicated.
+		if second.Index > highestIndex {
+			// No guarantee first segment has been replicated.
+			break
+		}
+
+		// Deletion can proceed!
+		size, err = first.Size()
+		if err != nil {
+			break
+		}
+		if err = os.Remove(first.Path); err != nil {
+			break
+		}
+		nBytesDeleted += size
+		segments = segments[1:]
+
+		// Create tombstone to indicate that the topic has been truncated at least once.
+		f, err := os.Create(t.tombstone)
```

Comment:
You can just call `return err` instead of `break` to make sure it's not shadowed.